x/stake/keeper/validator.go

```diff
@@ -201,16 +201,37 @@ func (k Keeper) GetValidatorsByPower(ctx sdk.Context) []types.Validator {
 // Accumulated updates to the active/bonded validator set for tendermint
 
 // get the most recently updated validators
+//
+// CONTRACT: Only validators with non-zero power or zero-power that were bonded
+// at the previous block height or were removed from the validator set entirely
+// are returned to Tendermint.
 func (k Keeper) GetTendermintUpdates(ctx sdk.Context) (updates []abci.Validator) {
 	store := ctx.KVStore(k.storeKey)
 
-	iterator := sdk.KVStorePrefixIterator(store, TendermintUpdatesKey) //smallest to largest
+	iterator := sdk.KVStorePrefixIterator(store, TendermintUpdatesKey)
 	for ; iterator.Valid(); iterator.Next() {
-		valBytes := iterator.Value()
-		var val abci.Validator
-		k.cdc.MustUnmarshalBinary(valBytes, &val)
-		updates = append(updates, val)
+		var abciVal abci.Validator
+
+		abciValBytes := iterator.Value()
+		k.cdc.MustUnmarshalBinary(abciValBytes, &abciVal)
+
+		val, found := k.GetValidator(ctx, abciVal.GetAddress())
+		if found {
+			// The validator is new or already exists in the store and must adhere to
+			// Tendermint invariants.
+			prevBonded := val.BondHeight < ctx.BlockHeight()
+			zeroPower := val.GetPower().Equal(sdk.ZeroDec())
+
```

Comment:
`val.GetPower()` doesn't mean the power which tendermint sees from EndBlock. `val.GetPower().RoundInt64()` represents the power which tendermint see. They have some difference. `val.GetPower()` returned data type is `sdk.Dec`. For instance, when `val.GetPower()` returns 0.4, then `val.GetPower().RoundInt64()` will be zero. Tendermint will remove the validator whose voting power is zero. However, staking module won't remove this validator.

Reply:
Yes @HaoyangLiu this issue/fix is unrelated to what you discovered. Thank you.